src/librustc_mir/build/expr/into.rs

```diff
@@ -179,20 +179,19 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         // conduct the test, if necessary
                         let body_block;
                         if let Some(cond_expr) = opt_cond_expr {
```

Comment:
N.B. As part of implementing `while a && let b = c { ... }` I'll make a PR to remove `hir::ExprKind::While` and then `hair::ExprKind::Loop` will presumably need to drop it's `condition` field and so the `else` branch would be floated out and the `if` branch would be removed... I hope the changes here do not cause problems for this...?

Reply:
That should be fine, this is only a problem because `while` loops are special cased. For any sensible HIR lowering this won't be a problem. The temporary is probably storage-live for the whole loop body, but that isn't observable. The following two functions generate almost identical optimized MIR after this PR, for example.

```rust
fn while_loop(c: bool) {
    while get_bool(c) {
        if get_bool(c) {
            break;
        }
    }
}

// What the above should probably expand to
fn exp_while_loop(c: bool) {
    loop {
        match get_bool(c) {
            true => {
                {if get_bool(c) {
                    break;
                }}
                continue;
            }
            _ => {}
        }
        break;
    }
}
```

Reply:
Cool.

> For any sensible HIR lowering this won't be a problem.

Specifically, the HIR lowering should likely be:

```rust
'label: while $cond $block
```

==>

```rust
'label: loop {
    match DropTemps($cond) {
        true => $block,
        _ => break,
    }
}
```

(is there a particular reason you are using `continue;` and an empty block?)

Reply:
Is slightly closer to what the RFC specified. What you're suggesting is probably better.

Reply:
This should be unnecessary now with https://github.com/rust-lang/rust/pull/61988.